src/python/pants/engine/process.py

```diff
@@ -290,9 +295,89 @@ def run(self, request: InteractiveProcess) -> InteractiveProcessResult:
         return self._scheduler.run_local_interactive_process(request)
 
 
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class BinaryPathRequest:
+    search_path: Tuple[str, ...]
+    binary_name: str
+
+    def __init__(self, *, search_path: Iterable[str], binary_name: str) -> None:
+        self.search_path = tuple(OrderedSet(search_path))
+        self.binary_name = binary_name
+
+
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class BinaryPaths(EngineAware):
+    binary_name: str
+    paths: Tuple[str, ...]
+
+    def __init__(self, binary_name: str, paths: Iterable[str]):
+        self.binary_name = binary_name
+        self.paths = tuple(OrderedSet(paths))
+
+    def level(self) -> Optional[LogLevel]:
+        return LogLevel.DEBUG if self.paths else LogLevel.WARN
+
+    def message(self) -> Optional[str]:
+        if not self.paths:
+            return f"failed to find {self.binary_name}"
+        found_msg = f"found {self.binary_name} at {self.paths[0]}"
+        if len(self.paths) > 1:
+            found_msg = f"{found_msg} and {pluralize(len(self.paths) - 1, 'other location')}"
+        return found_msg
+
+    @property
+    def first_path(self) -> Optional[str]:
+        """Return the first path to the binary that was discovered, if any."""
+        return next(iter(self.paths), None)
+
+
+@rule(desc="Find binary path")
+async def find_binary(request: BinaryPathRequest) -> BinaryPaths:
+    # TODO(John Sirois): Replace this script with a statically linked native binary so we don't
+    #  depend on either /bin/bash being available on the Process host.
+    # TODO: Once we stop using Google RBE, add a shebang to the script `/usr/bin/env bash`, rather
+    #  than running the binary `/bin/bash`. We hit a 26 "Text File Busy" error, which we suspect is
+    #  from RBE not having fsynced the ./script.sh file.
```

Comment:
I'll fix this TODO to point to https://github.com/pantsbuild/pants/issues/10507.